docs/src/spaces.md

```diff
@@ -0,0 +1,155 @@
+
+* Generated data that depends on the characteristics of the host system.  Examples are compiled binaries, fontcache system font folder inspection output, generated CUDA bitcode files, etc...  Objects that would be difficult to compute off of the user's machine, and that can be recreated without user intervention are a great fit.
+
+* Directories that should be shared between multiple packages in a single depot.  The space keying mechanism (explained above) makes it simple to provide scratch space that can be shared between different versions of a package, or even between different packages.  This allows packages to provide a scratch space where other packages can easily find the generated data, however the typical race condition warnings apply here; always design your access patterns assuming another process could be reading or writing to this scratch space at any time.
+
+Bad usecases for a Pkg scratchspace include (but are not limited to):
+
+* Anything that requires user input to regenerate.  Because spaces can disappear, it is a bad experience for the user to need to answer questions at seemingly random times when the space must be rebuilt.
+
+* Storing data that is write-once, read-many times.  We suggest you use [Artifacts](@ref) for that, as they are much more persistent and are built to become portable (so that other machines do not have to generate the data, they can simple make use of the artifact by downloading it from a hosted location).  Spaces generally should follow a write-many read-many access pattern.
+
+## Frequently-Accessed Caching Questions
+
+> Can I trigger data regeneration if the space is found to be empty/files are missing?
+
+Yes, this is quite simple; just check the contents of the directory when you first call `get_space!()`, and if it's empty, run your generation function:
+
+```julia
+using Pkg, Pkg.Spaces
+
+function get_dataset_dir()
+    dataset_dir = @get_space!("dataset")
+    if isempty(readdir(dataset_dir))
+        perform_expensive_dataset_generation(dataset_dir)
+    end
+    return dataset_dir
+end
+```
+
+> Can I create a scratchspace that is not shared across versions of my package?
+
+Yes!  Make use of the `key` parameter and Pkg's ability to look up the current version of your package at compile-time:
+
+```julia
+module VersionSpecificExample
+using Pkg, Pkg.Spaces
+
+# Helpers to get current package UUID and VersionNumber
+function get_uuid()
+    return Base.PkgId(@__MODULE__).uuid
+end
+function get_version(uuid)
+    ctx = Pkg.Types.Context()
+    uuid, entry = first(filter(((u, e),) -> u == uuid, ctx.env.manifest))
+    return entry.version
+end
+
+# Get the current version at compile-time, that's fine it's not going to change. ;)
+const v = get_version(get_uuid())
+
+# This will be filled in by `__init__()`; it might change if we get deployed somewhere
+version_specific_space = ""
```

Comment:
I wonder if it is in general better to advice people to use

```
version_specific_space = Ref{String}()
```

andb then 

```
version_specific_space[] = ...
```

in `__init__`?

That makes you able (for non isbitstypes) to actually check that the value has been set instead of not knowing whether you have the original `""` or it just got set to `""`.

Reply:
Hmmm, what situations are you thinking of that could allow for `version_specific_space` to _not_ be set?  It's within `__init__()`, so it should be guaranteed to either be set before any other (non-compile-time) code can run, or it will have thrown an error.